activesupport/lib/active_support/testing/assertions.rb

```diff
@@ -30,6 +30,71 @@ def assert_nothing_raised
         yield
       end
 
+      # Test varying numeric difference between the return value of an expression as a
+      # result of what is evaluated in the yielded block.
+      #
+      #   assert_differences ['Article.count'] do
+      #     post :create, params: { article: {...} }
+      #   end
+      #
+      # An arbitrary expression is passed in and evaluated.
+      #
+      #   assert_differences ['Article.last.comments(:reload).size'] do
+      #     post :create, params: { comment: {...} }
+      #   end
+      #
+      # An arbitrary positive or negative difference can be specified.
+      # The default is <tt>1</tt>.
+      #
+      #   assert_differences ['Article.count', -1] do
+      #     post :delete, params: { id: ... }
+      #   end
+      #
+      # An array of expressions and the predicted change can also be passed in and evaluated.
+      #
+      #   assert_differences [ ['Article.count', 2], ['Post.count', 1] ] do
+      #     post :create, params: { article: {...} }
+      #   end
+      #
+      # A lambda or a list of lambdas can be passed in and evaluated:
+      #
+      #   assert_differences [->{ Article.count }, 2] do
+      #     post :create, params: { article: {...} }
+      #   end
+      #
+      #   assert_differences [ [->{ Article.count }, 2], [->{ Post.count }, 1] ] do
+      #     post :create, params: { article: {...} }
+      #   end
+      #
+      # An error message can be specified.
+      #
+      #   assert_differences ['Article.count', -1], 'An Article should be destroyed' do
+      #     post :delete, params: { id: ... }
+      #   end
+      def assert_differences(expression_array, message = nil, &block)
+        expressions = Array(expression_array)
+        expressions = [expressions] unless expressions.all? { |e| e.kind_of? Array }
+
+        exps = expressions.map do |e|
+          callable = e.first.respond_to?(:call) ? e.first : lambda { eval(e.first, block.binding) }
+          {
+            callable: callable,
+            before: callable.call
+          }
+        end
```

Comment:
I don't love the idea of this hash return.  The idea being that assigning `before` here cuts the need for another loop, like what was present in the original `assert_difference`.

If no-one minds me having 2 loops, I can modify that.